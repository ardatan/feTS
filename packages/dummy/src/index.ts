'I am a dummy package';
