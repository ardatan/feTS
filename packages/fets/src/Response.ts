import { Response as OriginalResponse } from '@whatwg-node/fetch';
import { TypedResponseCtor } from './typed-fetch';
import { JSONSerializer } from './types';

export const LAZY_SERIALIZED_RESPONSE = Symbol('LAZY_SERIALIZED_RESPONSE');

export const defaultSerializer: JSONSerializer = obj => JSON.stringify(obj);

export interface LazySerializedResponse {
  [LAZY_SERIALIZED_RESPONSE]: true;
  resolveWithSerializer(serializer: JSONSerializer): void;
  init?: ResponseInit;
  serializerSet: boolean;
  responsePromise: Promise<Response>;
  jsonObj: any;
}

export function isLazySerializedResponse(response: any): response is LazySerializedResponse {
  return response != null && response[LAZY_SERIALIZED_RESPONSE];
}

export function createLazySerializedResponse(
  jsonObj: any,
  init?: ResponseInit,
): LazySerializedResponse {
  let resolve: (value: Response) => void;
  const promise = new Promise<Response>(_resolve => {
    resolve = _resolve;
  });
  let _serializerSet = false;
  return {
    jsonObj,
    responsePromise: promise,
    [LAZY_SERIALIZED_RESPONSE]: true,
    init,
    get serializerSet() {
      return _serializerSet;
    },
    resolveWithSerializer(serializer: JSONSerializer) {
      const serialized = serializer(jsonObj);
      _serializerSet = true;
      resolve(
        new OriginalResponse(serialized, {
          ...init,
          status: init?.status || 200,
          headers: {
            ...init?.headers,
            'Content-Type': 'application/json',
          },
        }),
      );
    },
  };
}

// This allows us to hook into serialization of the response body
export const Response = new Proxy(OriginalResponse, {
  get(OriginalResponse, prop, receiver) {
    if (prop === 'json') {
      return createLazySerializedResponse;
    }
    return Reflect.get(OriginalResponse, prop, receiver);
  },
}) as TypedResponseCtor;
