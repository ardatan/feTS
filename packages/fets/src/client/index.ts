export * from './createClient';
export * from './types';
