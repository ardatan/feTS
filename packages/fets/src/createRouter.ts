import { parse as qsParse } from 'qs';
import * as DefaultFetchAPI from '@whatwg-node/fetch';
import { createServerAdapter, isPromise, useErrorHandling } from '@whatwg-node/server';
import landingPageRaw from './landing-page.js';
import { useDefineRoutes } from './plugins/define-routes.js';
import { useOpenAPI } from './plugins/openapi.js';
import { useTypeBox } from './plugins/typebox.js';
import { HTTPMethod, TypedRequest, TypedResponse } from './typed-fetch.js';
import type {
  OnRouteHandleHook,
  OnRouteHook,
  OnRouterInitHook,
  OpenAPIDocument,
  OpenAPIInfo,
  Router,
  RouterBaseObject,
  RouterComponentsBase,
  RouterOptions,
  RouterPlugin,
  RouterSDK,
  RouteSchemas,
  RouteWithSchemasOpts,
} from './types.js';
import { asyncIterationUntilReturn } from './utils.js';

const EMPTY_OBJECT = {};

export function createRouterBase(
  {
    fetchAPI: givenFetchAPI,
    base: basePath = '/',
    plugins = [],
    openAPI,
    swaggerUI,
    landingPage = true,
  }: RouterOptions<any, any> = {},
  openAPIDocument: OpenAPIDocument,
): RouterBaseObject<any, any, any> {
  const fetchAPI = {
    ...DefaultFetchAPI,
    ...givenFetchAPI,
  };
  const __onRouterInitHooks: OnRouterInitHook<any>[] = [];
  const onRouteHooks: OnRouteHook<any>[] = [];
  const onRouteHandleHooks: OnRouteHandleHook<any, RouterComponentsBase>[] = [];
  for (const plugin of plugins) {
    if (plugin.onRouterInit) {
      __onRouterInitHooks.push(plugin.onRouterInit);
    }
    if (plugin.onRoute) {
      onRouteHooks.push(plugin.onRoute);
    }
    if (plugin.onRouteHandle) {
      onRouteHandleHooks.push(plugin.onRouteHandle);
    }
  }
  const routeByPatternByMethod = new Map<
    HTTPMethod,
    Map<
      URLPattern,
      RouteWithSchemasOpts<
        any,
        RouterComponentsBase,
        RouteSchemas,
        HTTPMethod,
        string,
        TypedRequest,
        TypedResponse
      >
    >
  >();
  const routeByPathByMethod = new Map<
    HTTPMethod,
    Map<
      string,
      RouteWithSchemasOpts<
        any,
        RouterComponentsBase,
        RouteSchemas,
        HTTPMethod,
        string,
        TypedRequest,
        TypedResponse
      >
    >
  >();

  function handleUnhandledRoute(requestPath: string) {
    if (landingPage) {
      return new fetchAPI.Response(
        landingPageRaw
          .replaceAll('__BASE_PATH__', basePath)
          .replaceAll('__OAS_PATH__', openAPI?.endpoint || '/openapi.json')
          .replaceAll('__SWAGGER_UI_PATH__', swaggerUI?.endpoint || '/docs')
          .replaceAll('__REQUEST_PATH__', requestPath),
        {
          headers: {
            'Content-Type': 'text/html',
          },
        },
      );
    }
    return undefined as any;
  }

  return {
    openAPIDocument,
    handle(request: Request, context: any) {
      let url = new Proxy(EMPTY_OBJECT as URL, {
        get(_target, prop, _receiver) {
          url = new fetchAPI.URL(request.url, 'http://localhost');
          return Reflect.get(url, prop, url);
        },
      }) as URL;
      let parsedQuery: any;
      Object.defineProperties(request, {
        parsedUrl: {
          get() {
            return url;
          },
          configurable: true,
        },
        query: {
          get() {
            if (parsedQuery == null) {
              const searchPart = request.url.split('?')[1];
              if (searchPart == null) {
                return EMPTY_OBJECT;
              }
              parsedQuery = qsParse(searchPart);
            }
            return parsedQuery;
          },
          configurable: true,
        },
      });
      const pathPatternMapByMethod = routeByPathByMethod.get(request.method as HTTPMethod);
      if (pathPatternMapByMethod) {
        const route = pathPatternMapByMethod.get(url.pathname);
        if (route) {
          for (const onRouteHandleHook of onRouteHandleHooks) {
            onRouteHandleHook({
              route,
              request: request as TypedRequest,
            });
          }
          const handlerResult$ = route.handler(request as TypedRequest, context);
          if (isPromise(handlerResult$)) {
            return handlerResult$.then(handlerResult => {
              if (handlerResult) {
                return handlerResult as Response;
              }
              return handleUnhandledRoute(request.url);
            });
          }
          if (handlerResult$) {
            return handlerResult$ as Response;
          }
        }
      }
      const methodPatternMaps = routeByPatternByMethod.get(request.method as HTTPMethod);
      if (methodPatternMaps) {
        const patternHandlerResult$ = asyncIterationUntilReturn(
          methodPatternMaps.entries(),
          ([pattern, route]) => {
            // Do not parse URL if not needed
            const match = pattern.exec(url);
            if (match != null) {
              let paramsProxy: any;
              Object.defineProperty(request, 'params', {
                get() {
                  if (paramsProxy == null) {
                    paramsProxy = new Proxy(match!.pathname.groups, {
                      get(_, prop) {
                        const value = (match!.pathname.groups as Record<string, string>)[
                          prop.toString()
                        ];
                        if (value != null) {
                          return decodeURIComponent(value);
                        }
                        return value;
                      },
                    });
                  }
                  return paramsProxy;
                },
                configurable: true,
              });
              for (const onRouteHandleHook of onRouteHandleHooks) {
                onRouteHandleHook({
                  route,
                  request: request as TypedRequest,
                });
              }
              return route.handler(request as TypedRequest, context);
            }
          },
        );
        if (isPromise(patternHandlerResult$)) {
          return patternHandlerResult$.then(patternHandlerResult => {
            if (patternHandlerResult) {
              return patternHandlerResult as Response;
            }
            return handleUnhandledRoute(request.url);
          });
        }
        if (patternHandlerResult$) {
          return patternHandlerResult$ as Response;
        }
      }
      return handleUnhandledRoute(request.url);
    },
    route(
      route: RouteWithSchemasOpts<
        any,
        RouterComponentsBase,
        RouteSchemas,
        HTTPMethod,
        string,
        TypedRequest,
        TypedResponse
      >,
    ) {
      for (const onRouteHook of onRouteHooks) {
        onRouteHook({
          basePath,
          route,
          routeByPathByMethod,
          routeByPatternByMethod,
          openAPIDocument,
          fetchAPI,
        });
      }
      return this as any;
    },
    __client: {},
    __onRouterInitHooks,
  };
}

export function createRouter<
  TServerContext,
  const TComponents extends RouterComponentsBase = {},
  const TRouterSDK extends RouterSDK<string, TypedRequest, TypedResponse> = {
    [TKey: string]: never;
  },
>(
  options: RouterOptions<TServerContext, TComponents> = {},
): Router<TServerContext, TComponents, TRouterSDK> {
  const {
    openAPI: {
      endpoint: oasEndpoint = '/openapi.json',
      includeValidationErrors,
      ...openAPIDocument
    } = {},
    swaggerUI: { endpoint: swaggerUIEndpoint = '/docs', ...swaggerUIOpts } = {},
    plugins: userPlugins = [],
    base = '/',
  } = options;
  openAPIDocument.openapi = openAPIDocument.openapi || '3.0.1';
  const oasInfo = (openAPIDocument.info ||= {} as OpenAPIInfo);
  oasInfo.title ||= 'feTS API';
  oasInfo.description ||= 'An API written with feTS';
  oasInfo.version ||= '1.0.0';
  if (base !== '/') {
    openAPIDocument.servers = openAPIDocument.servers || [
      {
        url: base,
      },
    ];
  }
  const plugins: RouterPlugin<TServerContext, TComponents>[] = [
    ...(oasEndpoint || swaggerUIEndpoint
      ? [
          useOpenAPI<TServerContext, TComponents>({
            oasEndpoint,
            swaggerUIEndpoint,
            swaggerUIOpts,
            includeValidationErrors,
          }),
        ]
      : []),
    useTypeBox(),
    useErrorHandling(options?.onError),
    useDefineRoutes(),
    ...userPlugins,
  ];
  const finalOpts: RouterOptions<TServerContext, TComponents> = {
    ...options,
    swaggerUI: {
      endpoint: swaggerUIEndpoint,
      ...swaggerUIOpts,
    },
    base,
    plugins,
  };
  const routerBaseObject = createRouterBase(finalOpts, openAPIDocument as OpenAPIDocument);
  const router = createServerAdapter(routerBaseObject, finalOpts);
  for (const onRouterInitHook of routerBaseObject.__onRouterInitHooks) {
    onRouterInitHook(router);
  }
  return router;
}
