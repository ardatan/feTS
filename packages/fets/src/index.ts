export * from './types.js';
export * from './createRouter.js';
export { URLPattern } from '@whatwg-node/fetch';
export { useCORS, useErrorHandling, HTTPError } from '@whatwg-node/server';
export { FromSchema } from 'json-schema-to-ts';
export * from './client/index.js';
export * from './Response.js';
export { useAjv } from './plugins/ajv.js';
