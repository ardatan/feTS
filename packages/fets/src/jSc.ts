import { JSONSchema } from './types';

type RequiredPropertyKeys<TProperties extends Record<string, JSONSchema>> = {
  [K in keyof TProperties]: TProperties[K] extends { ['__optional']: true } ? never : K;
}[keyof TProperties];

const sharedHelpers = {
  $format<TThis extends JSONSchema, TFormat extends string>(
    this: TThis,
    format: TFormat,
  ): Omit<TThis, 'format'> & { ['format']: TFormat } {
    return {
      ...this,
      format,
    } as const;
  },
  $enum<TThis extends JSONSchema, TEnums extends string[]>(
    this: TThis,
    ..._enum: TEnums
  ): Omit<TThis, 'enum'> & { enum: TEnums } {
    return {
      ...this,
      enum: _enum,
    } as const;
  },
  $nullable<TThis extends JSONSchema, TNullable extends boolean = true>(
    this: TThis,
    nullable: TNullable = true as TNullable,
  ): Omit<TThis, 'nullable'> & { ['nullable']: TNullable } {
    return {
      ...this,
      nullable,
    } as const;
  },
  $optional<TThis extends JSONSchema, TOptional extends boolean = true>(
    this: TThis,
    __optional: TOptional = true as TOptional,
  ): Omit<TThis, '__optional'> & { ['__optional']: TOptional } {
    return {
      ...this,
      __optional,
    } as const;
  },
  $not<TThis extends JSONSchema, TNot extends JSONSchema>(
    this: TThis,
    not: TNot,
  ): Omit<TThis, 'not'> & { ['not']: TNot } {
    return {
      ...this,
      not,
    } as const;
  },
} as const;

export const jSc = {
  $string() {
    return {
      type: 'string',
      $maxLength<TThis extends JSONSchema, TMaxLength extends number>(
        this: TThis,
        maxLength: TMaxLength,
      ): Omit<TThis, 'maxLength'> & { ['maxLength']: TMaxLength } {
        return {
          ...this,
          maxLength,
        } as const;
      },
      $minLength<TThis extends JSONSchema, TMinLength extends number>(
        this: TThis,
        minLength: TMinLength,
      ): Omit<TThis, 'minLength'> & { ['minLength']: TMinLength } {
        return {
          ...this,
          minLength,
        } as const;
      },
      $length<TThis extends JSONSchema, TLength extends number>(
        this: TThis,
        length: TLength,
      ): Omit<TThis, 'minLength' | 'maxLength'> & {
        ['minLength']: TLength;
        ['maxLength']: TLength;
      } {
        return {
          ...this,
          minLength: length,
          maxLength: length,
        } as const;
      },
      $pattern<TPattern extends string>(pattern: TPattern) {
        return {
          ...this,
          pattern,
        } as const;
      },
      ...sharedHelpers,
    } as const;
  },
  $number() {
    return {
      type: 'number',
      $min<TMin extends number>(min: TMin) {
        return {
          ...this,
          minimum: min,
        } as const;
      },
      $max<TMax extends number>(max: TMax) {
        return {
          ...this,
          maximum: max,
        } as const;
      },
      ...sharedHelpers,
    } as const;
  },
  $boolean() {
    return {
      type: 'boolean',
      ...sharedHelpers,
    } as const;
  },
  $null() {
    return {
      type: 'null',
      ...sharedHelpers,
    } as const;
  },
  $object<TProperties extends Record<string, JSONSchema>>(properties: TProperties) {
    const required = [] as RequiredPropertyKeys<TProperties>[];
    for (const key in properties) {
      if ((properties[key] as any).__optional) {
        continue;
      }
      required.push(key as unknown as RequiredPropertyKeys<TProperties>);
    }
    return {
      type: 'object',
      properties,
      additionalProperties: false,
      required,
      $additionalProperties<TThis extends JSONSchema, TAdditionalProperties extends JSONSchema>(
        this: TThis,
        additionalProperties: TAdditionalProperties,
      ): Omit<TThis, 'additionalProperties'> & { ['additionalProperties']: TAdditionalProperties } {
        return {
          ...this,
          additionalProperties,
        } as const;
      },
      ...sharedHelpers,
    } as const;
  },
  $array<TItems extends JSONSchema>(items: TItems) {
    return {
      type: 'array',
      items,
      $minItems<TThis extends JSONSchema, TMinItems extends number>(
        this: TThis,
        minItems: TMinItems,
      ): Omit<TThis, 'minItems'> & { ['minItems']: TMinItems } {
        return {
          ...this,
          minItems,
        } as const;
      },
      $maxItems<TThis extends JSONSchema, TMaxItems extends number>(
        this: TThis,
        maxItems: TMaxItems,
      ): Omit<TThis, 'maxItems'> & { ['maxItems']: TMaxItems } {
        return {
          ...this,
          maxItems,
        } as const;
      },
      $itemsLength<TThis extends JSONSchema, TItemsLength extends number>(
        this: TThis,
        itemsLength: TItemsLength,
      ): Omit<TThis, 'minItems' | 'maxItems'> & {
        ['minItems']: TItemsLength;
        ['maxItems']: TItemsLength;
      } {
        return {
          ...this,
          minItems: itemsLength,
          maxItems: itemsLength,
        } as const;
      },
      ...sharedHelpers,
    } as const;
  },
  $oneOf<TOneOf extends JSONSchema[]>(...oneOf: TOneOf) {
    return {
      oneOf,
    } as const;
  },
  $anyOf<TAnyOf extends JSONSchema[]>(...anyOf: TAnyOf) {
    return {
      anyOf,
    } as const;
  },
  $allOf<TAllOf extends JSONSchema[]>(...allOf: TAllOf) {
    return {
      allOf,
    } as const;
  },
};
