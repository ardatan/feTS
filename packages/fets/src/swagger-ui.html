<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="description" content="SwaggerUI" />
    <title>SwaggerUI</title>
    <style>
      @import 'https://raw.githubusercontent.com/Itz-fork/Fastapi-Swagger-UI-Dark/main/assets/swagger_ui_dark.min.css'
        (prefers-color-scheme: dark);

      @import 'https://unpkg.com/swagger-ui-dist/swagger-ui.css' (prefers-color-scheme: light);

      @media (prefers-color-scheme: dark) {
        .swagger-ui textarea {
          background: hsl(358.8deg 1.05% 1.91% / 80%) !important;
        }
        .swagger-ui .scheme-container {
          background: #1e1e2e;
        }
        .swagger-ui select {
          background: rgba(0, 0, 0, 0.2);
        }
      }
    </style>
  </head>

  <body>
    <div id="swagger-ui"></div>
    <script src="https://unpkg.com/swagger-ui-dist/swagger-ui-bundle.js" crossorigin></script>
    <script>
      window.onload = () => {
        window.ui = SwaggerUIBundle(__SWAGGER_UI_OPTIONS__);
      };
    </script>
  </body>
</html>
