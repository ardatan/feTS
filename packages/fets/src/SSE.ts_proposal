import { TransformStream } from "@whatwg-node/fetch";
import { defaultSerializer, Response } from "./Response";
import { SSEBody, TypedResponse, TypedResponseInit } from "./typed-fetch";
import { JSONSerializer } from "./types"

export type SSEMessage<TEventDataMap extends Record<string, any>> = {
    [TEventName in keyof TEventDataMap]: {
        event: TEventName;
        data: TEventDataMap[TEventName];
    }
}[keyof TEventDataMap];

export class SSEMessageToByteStream<TEventDataMap extends Record<string, any>> extends TransformStream<SSEMessage<TEventDataMap>, Uint8Array> {
    public serializer: JSONSerializer = defaultSerializer;
    private textEncoder = new TextEncoder();
    constructor() {
        super({
            transform: (chunk, controller) => {
                const {
                    event,
                    data,
                } = chunk;
                const serializedData = this.serializer(data);
                const sseEncodedMessage = this.textEncoder.encode(`event: ${event.toString()}\ndata: ${serializedData}\n\n`);
                controller.enqueue(sseEncodedMessage);
            }
        })
    }
}

export class ByteToSSEMessageStream<TEventDataMap extends Record<string, any>> extends TransformStream<Uint8Array, SSEMessage<TEventDataMap>> {
    public serializer: JSONSerializer = defaultSerializer;
    private textDecoder = new TextDecoder();
    constructor() {
        super({
            transform: (chunk, controller) => {
                const decodedMessage = this.textDecoder.decode(chunk);
                const [event, dataStr] = decodedMessage.split("data: ");
                const parsedData = JSON.parse(dataStr);
                controller.enqueue({
                    event,
                    data: parsedData,
                });
            }
        })
    }
}

export function readableStreamToAsyncIterable<T>(readableStream: ReadableStream<T>): AsyncIterable<T> {
    if (Symbol.asyncIterator in readableStream) {
        return readableStream as AsyncIterable<T>;
    }
    return {
        [Symbol.asyncIterator]() {
            const reader = readableStream.getReader();
            return {
                next() {
                    return reader.read();
                },
                return() {
                    reader.releaseLock();
                    return {};
                },
                async throw(error) {
                    await reader.cancel(error);
                    return {
                        done: true,
                    }
                }
            } as AsyncIterator<T>;
        }
    }
}

export function getSSEAsyncIterableFromBody<TEventDataMap extends Record<string, any>>(body: SSEBody<TEventDataMap>): AsyncIterable<SSEMessage<TEventDataMap>> {
    const byteToSSEMessageStream = new ByteToSSEMessageStream<TEventDataMap>();
    const readableStream = body.pipeThrough(byteToSSEMessageStream);
    return readableStreamToAsyncIterable(readableStream);
}

export function createSSEResponse<TEventDataMap extends Record<string, any>, TStatusCode extends number>(init?: TypedResponseInit<TStatusCode>): {
   writer: WritableStreamDefaultWriter<SSEMessage<TEventDataMap>>;
   response: TypedResponse<any, TEventDataMap>;
} {
    const transformStream = new SSEMessageToByteStream<TEventDataMap>();
    const response = new Response(transformStream.readable, {
        ...init,
        headers: {
            'Content-Type': 'text/event-stream',
            ...init?.headers,
        },
        status: init?.status || 200,
    }) as TypedResponse<any, TEventDataMap>;
    return {
        writer: transformStream.writable.getWriter(),
        response,
    };
}


export type SSEBody<TEventDataMap extends Record<string, any>> = ReadableStream<Uint8Array> & {
    __eventDataMap?: TEventDataMap;
}