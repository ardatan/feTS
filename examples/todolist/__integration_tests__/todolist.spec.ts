/* eslint-disable camelcase */
import { globalAgent } from 'http';
import {
  us_listen_socket_close,
  us_socket_local_port,
  type us_listen_socket,
} from 'uWebSockets.js';
import { fetch } from '@whatwg-node/fetch';
import { app } from '../src/app';

describe('TodoList', () => {
  const nodeMajor = parseInt(process.versions.node.split('.')[0], 10);
  if (nodeMajor < 16) {
    it('should be skipped', () => {});
    return;
  }
  let listenSocket: us_listen_socket;
  let port: number;
  beforeAll(done => {
    app.listen(0, newListenSocket => {
      listenSocket = newListenSocket;
      if (!listenSocket) {
        done.fail('Failed to start the server');
        return;
      }
      port = us_socket_local_port(listenSocket);
      done();
    });
  });
  afterAll(() => {
    if (listenSocket) {
      us_listen_socket_close(listenSocket);
    }
    globalAgent.destroy();
  });
  it('should work', async () => {
    const response = await fetch(`http://localhost:${port}/todos`);
    expect(response.status).toBe(200);
    expect(await response.json()).toEqual([]);
  });
  it('should show Swagger UI', async () => {
    const response = await fetch(`http://localhost:${port}/docs`);
    expect(response.status).toBe(200);
    expect(response.headers.get('content-type')).toBe('text/html');
    const html = await response.text();
    expect(html).toContain('<title>SwaggerUI</title>');
  });
  it('should expose OpenAPI document', async () => {
    const response = await fetch(`http://localhost:${port}/openapi.json`);
    expect(response.status).toBe(200);
    expect(response.headers.get('content-type')).toContain('application/json');
    const json = await response.json();
    expect(json).toMatchSnapshot();
  });
});
