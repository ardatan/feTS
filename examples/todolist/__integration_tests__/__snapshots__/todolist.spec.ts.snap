// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should expose OpenAPI document 1`] = `
{
  "components": {
    "schemas": {
      "Todo": {
        "additionalProperties": false,
        "properties": {
          "content": {
            "type": "string",
          },
          "id": {
            "type": "string",
          },
        },
        "required": [
          "id",
          "content",
        ],
        "type": "object",
      },
    },
  },
  "info": {
    "description": "A simple todo list example with feTS",
    "title": "Todo List Example",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": {
    "/todo": {
      "put": {
        "description": "Add a todo",
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "additionalProperties": false,
                "properties": {
                  "content": {
                    "type": "string",
                  },
                },
                "required": [
                  "content",
                ],
                "type": "object",
              },
            },
          },
          "required": true,
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/Todo",
                },
              },
            },
            "description": "",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "message": {
                      "type": "string",
                    },
                    "name": {
                      "type": "string",
                    },
                    "path": {
                      "type": "string",
                    },
                    "value": {
                      "anyOf": [
                        {
                          "type": "string",
                        },
                        {
                          "type": "number",
                        },
                        {
                          "type": "boolean",
                        },
                        {
                          "type": "array",
                        },
                        {
                          "type": "object",
                        },
                      ],
                    },
                  },
                  "title": "RequestValidationError",
                  "type": "object",
                },
              },
            },
            "description": "Request validation failed",
          },
        },
      },
    },
    "/todo/{id}": {
      "delete": {
        "description": "Delete a todo",
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "id": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "id",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "message": {
                      "type": "string",
                    },
                    "name": {
                      "type": "string",
                    },
                    "path": {
                      "type": "string",
                    },
                    "value": {
                      "anyOf": [
                        {
                          "type": "string",
                        },
                        {
                          "type": "number",
                        },
                        {
                          "type": "boolean",
                        },
                        {
                          "type": "array",
                        },
                        {
                          "type": "object",
                        },
                      ],
                    },
                  },
                  "title": "RequestValidationError",
                  "type": "object",
                },
              },
            },
            "description": "Request validation failed",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "error": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "error",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
      "get": {
        "description": "Get a todo",
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/Todo",
                },
              },
            },
            "description": "",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "message": {
                      "type": "string",
                    },
                    "name": {
                      "type": "string",
                    },
                    "path": {
                      "type": "string",
                    },
                    "value": {
                      "anyOf": [
                        {
                          "type": "string",
                        },
                        {
                          "type": "number",
                        },
                        {
                          "type": "boolean",
                        },
                        {
                          "type": "array",
                        },
                        {
                          "type": "object",
                        },
                      ],
                    },
                  },
                  "title": "RequestValidationError",
                  "type": "object",
                },
              },
            },
            "description": "Request validation failed",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "message": {
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "",
          },
        },
      },
    },
    "/todos": {
      "get": {
        "description": "Get all todos",
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "items": {
                    "$ref": "#/components/schemas/Todo",
                  },
                  "type": "array",
                },
              },
            },
            "description": "",
          },
        },
      },
    },
    "/upload": {
      "post": {
        "description": "Upload a file",
        "requestBody": {
          "content": {
            "multipart/form-data": {
              "schema": {
                "additionalProperties": false,
                "properties": {
                  "description": {
                    "maxLength": 255,
                    "type": "string",
                  },
                  "file": {
                    "format": "binary",
                    "type": "string",
                  },
                },
                "required": [
                  "file",
                ],
                "type": "object",
              },
            },
          },
          "required": true,
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "description": {
                      "type": "string",
                    },
                    "lastModified": {
                      "type": "number",
                    },
                    "name": {
                      "type": "string",
                    },
                    "size": {
                      "type": "number",
                    },
                    "type": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "name",
                    "type",
                    "size",
                    "lastModified",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "message": {
                      "type": "string",
                    },
                    "name": {
                      "type": "string",
                    },
                    "path": {
                      "type": "string",
                    },
                    "value": {
                      "anyOf": [
                        {
                          "type": "string",
                        },
                        {
                          "type": "number",
                        },
                        {
                          "type": "boolean",
                        },
                        {
                          "type": "array",
                        },
                        {
                          "type": "object",
                        },
                      ],
                    },
                  },
                  "title": "RequestValidationError",
                  "type": "object",
                },
              },
            },
            "description": "Request validation failed",
          },
        },
      },
    },
  },
}
`;
